### PR TITLE
Added back GQL for enabling developer preview, updated renderDev footer to include toggle for developer preview

### PR DESCRIPTION
**Pull request by jamieguerrero**
### WHY are these changes introduced?

Closes https://github.com/Shopify/internal-cli-foundations/issues/713

Implements solution agreed upon in this [doc discussing Dev Preview](https://docs.google.com/document/d/1pJiB-ZWFP7ieKTVAhmC6hEtI2ZuIoNGH7iFhqO-zhXA/edit?usp=sharing).

IMPORTANT!!!! This [PR](https://github.com/Shopify/cli/pull/2637) should be merged or rejected first. Other alternative is rebase this branch to the PR's one and then merge this PR

### WHAT is this pull request doing?

**Previously**, when running `shopify app dev`, if you have at least one `draftable` extension you are prompted to turn on Developer Preview in your store via the green underlined link to "Partner Dashboard".

Screenshot 2023-08-03 at 4 03 26 PM

**After this PR,** when running `shopify app dev`, it will automatically set dev preview to 'on' and render another shortcut in the footer that will allow you to toggle the developer preview mode with `d`. 

### How to test your changes?

- Clone Shopify/cli
- Checkout the `add-dev-preview-toggle` branch
- Create an app using `pnpm create-app`
- cd into app folder `cd ./name-of-app`
- Generate a draftable extension like admin-action or admin-block with `pnpm shopify app generate extension`
- cd into cli folder `cd ..`
- Run app dev command on newly created app with `pnpm shopify app dev --path ./name-of-app`
- Install app on your development store
- You should see the new toggle for turning off/on developer preview 🎉 

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Comment by jessehoek** [comment]
Thanks for working on this @jamieguerrero! A language note: internally for brevity a lot of us call this "dev preview". But the setting is specifically called "development store preview" in the Dashboard and docs. 

We also don't use the term "enable" and "disable" in the dashboard. We say "turn on" and "turn off". In this case, could the shortcut always be "d" to stand for "development store preview"? That follows the pattern for the below 2 key shortcuts.

So basically that would look like
`Press d  |  development store preview: on|off`

File: packages/app/src/cli/services/dev/output.ts [hunk 1]
```diff
@@ -76,7 +76,11 @@ export function renderDev(
   const shortcuts = []
   const enabledStorePreviewShortcut = developmentStorePreviewEnabled !== undefined
   if (enabledStorePreviewShortcut) {
-    shortcuts.push(buildDevPreviewShortcut(developmentStorePreviewEnabled, apiKey, token))
+    // Enable dev preview on app dev start
+    await developerPreviewUpdate(apiKey, token, true)
+    buildDevPreviewShortcut(true, apiKey, token)
+
+    shortcuts.push(buildDevPreviewShortcut(true, apiKey, token))
```
Comment (alvaro-shopify) [inline packages/app/src/cli/services/dev/output.ts:83]:
```ts
const result = await developerPreviewUpdate(apiKey, token, true)
shortcuts.push(buildDevPreviewShortcut(result, apiKey, token))
```

I would add an utility method `enableDeveloperPreview` to avoid using `true` or `false` in cases where we know in advance the action to execute

File: packages/app/src/cli/services/dev.ts [hunk 2]
```diff
@@ -329,19 +328,29 @@ async function dev(options: DevOptions) {
 
   await reportAnalyticsEvent({config: options.commandConfig})
 
+  const enabledPreviewMode = isPreviewModeEnabled(remoteApp, localApp) ? developmentStorePreviewEnabled : undefined
```
[the remark this answers is not in the document]
Comment (matteodepalo) [reply packages/app/src/cli/services/dev.ts:331]: Even in that scenario though I would recommend against using `undefined` to represent state. What if instead we passed both `canEnablePreviewMode` and `isPreviewModeEnabled` inside `app` and let the user of this param decide wether to use the value provided or not?

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 3]
```diff
@@ -12,19 +12,31 @@ import {Writable} from 'stream'
...
 export interface ConcurrentOutputProps {
   processes: OutputProcess[]
   abortSignal: AbortSignal
   showTimestamps?: boolean
-  onInput?: (input: string, key: Key, exit: () => void) => void
-  footer?: {
-    shortcuts: Shortcut[]
-    subTitle?: string
-  }
+  onInput?: (input: string, key: Key, exit: () => void, footerContext: FooterContext) => Promise<void>
```
Comment (matteodepalo) [inline packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:38]: This is breaking change so we should make sure that none is using `onInput` at Shopify or they will get compilation errors.

File: packages/app/src/cli/services/dev.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamieguerrero) [reply packages/app/src/cli/services/dev.ts:331]: Great suggestion on renaming these variables! Naming is so hard, eh? 😅 

Also, passed both canEnablePreviewMode and developmentStorePreviewEnabled into renderDev, so we can avoid using undefined to represent state. Good call!

File: packages/app/src/cli/services/dev/output.ts [hunk 4]
```diff
@@ -98,3 +148,26 @@ async function partnersURL(organizationId: string, appId: string) {
...
 }
+
+function buildPollForDevPreviewMode(apiKey: string, token: string, interval = 10) {
+  return (footerContext: FooterContext) => {
+    const currentIntervalInSeconds = interval
+
+    return new Promise<void>((_resolve, _reject) => {
+      const onPoll = async () => {
+        const app = await fetchAppFromApiKey(apiKey, token)
+        const currentShortcutAction = footerContext.footer?.shortcuts.find((shortcut) => shortcut.key === 'd')
```
[the remark this answers is not in the document]
Comment (jamieguerrero) [reply packages/app/src/cli/services/dev/output.ts:159]: Screenshot 2023-08-11 at 10 43 04 AM

We do need this check because if for any reason the shortcut doesn't exist, we should do nothing. The updateShortcut function will also complain if it receives undefined.

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx (hunk 3 above)
Comment (jamieguerrero) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx]: Great call! We have made footerContext optional, that should make onInput backwards compatible.

Comment (jamieguerrero) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx]: After reconsideration, we made onInputAsync as a separate optional handler in case folks want to use async methods to handle their shortcuts. This should not be a breaking change!

File: packages/app/src/cli/services/extensions/common.ts [hunk 5]
```diff
@@ -49,28 +47,13 @@ export async function ensureExtensionDirectoryExists({name, app}: {name: string;
...
-  if (!body) return
-
-  outputInfo(`\n${body}\n`)
-}
-
-export async function buildDevPreviewWarning(remoteApp: Partial<OrganizationApp>, localApp: AppInterface) {
+export function isPreviewModeEnabled(remoteApp: Partial<OrganizationApp>, localApp: AppInterface) {
   const unifiedDeployment = remoteApp?.betas?.unifiedAppDeployment ?? false
-  if (!unifiedDeployment) return
+  if (!unifiedDeployment) return false
```
[the remark this answers is not in the document]
Comment (matteodepalo) [reply packages/app/src/cli/services/extensions/common.ts:52]: Let's ignore this suggestion for now. Once we go through the codebase to remove code related to unified deployments we can remove this check too.

**Review by matteodepalo — APPROVED** [review]
Nice!

I'm not too worried about making many customization to renderConcurrent. At some point we're going to add `ink` to `app` directly so that we can create custom components like `<Dev>` and move this custom stuff there.

File: packages/app/src/cli/services/extensions/common.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/services/extensions/common.ts:52]: @matteodepalo what we are doing here is not displaying the `dev preview` toggle shortcut in the footer in case `unified deployments` is not enabled. I think that in that case the `dev preview mode` does not take any effect on how devs preview their changes right?. `q` and `p` shortcuts are always available regardless the beta flag

[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/services/extensions/common.ts:52]: But in case you don't opt into `unified deployments` and you have a `theme app extension` I don't know if the new API request will work. I will check that and came back to you with the feedback

[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/services/extensions/common.ts:52]: @matteodepalo you are totally right!, it's possible to use the API and add the new shortcut so it's easier for app with theme app extension that has not opted into `unified deployments`

**Comment by jessehoek** [comment]
That looks great, just tophatted and everything works as expected. 🎉 UX looks good to me.
